### DISCUSSION TITLE
Unable to reset values when form fields are dynamically set

### DISCUSSION THREAD

**Discussion by spidertwin2** [react-hook-form/react-hook-form#5797] [Q&A] [answered]
Aloha :sun_with_face: 

I have a Form higher order component whose children become form fields.  When using `reset` inside `useEffect`, it is not clearing the form fields, i.e. still showing user entered content.  What is the right way to use `reset` when the field values are dynamically set?

Form.tsx
```
export function Form<S extends z.ZodType<any, any>>({
  schema,
  children,
  onSubmit,
  submitText,
  initialValues,
  ...props
}: FormProps<S>) {
  const ctx = useForm<z.infer<S>>({
    mode: "onBlur",
    resolver: async (values) => {
      try {
        if (schema) {
          schema.parse(values)
        }
        return { values, errors: {} }
      } catch (error) {
        return { values: {}, errors: error.formErrors?.fieldErrors }
      }
    },
    defaultValues: initialValues,
  })

  const [formError, setFormError] = useState<string | null>(null)

  useEffect(() => {
    if (ctx.formState.isSubmitSuccessful) {
      ctx.reset() // this doesn't reset the field values
    }
  }, [ctx.reset, ctx.formState, ctx])

  return (
    <FormProvider {...ctx}>
      <form
        onSubmit={ctx.handleSubmit(async (values) => {
          const result = (await onSubmit(values)) || {}
          for (const [key, value] of Object.entries(result)) {
            if (key === FORM_ERROR) {
              setFormError(value)
            } else {
              ctx.setError(key as any, {
                type: "submit",
                message: value,
              })
            }
          }
        })}
        className="form"
        {...props}
      >
        {/* Form fields supplied as children are rendered here */}
        {children}

        {formError && console.log(formError)}

          <Button
            type="submit"
            loadingText="Submitting"
            isLoading={ctx.formState.isSubmitting}
            disabled={ctx.formState.isSubmitting}
            {...submitButtonProps}
          >
            {submitText}
          </Button>
      
      </form>
    </FormProvider>
  )
}
```

**Comment by thanh-nguyen-95** [comment]
Hi, can you provide a CodeSandbox to reproduce the problem? It doesn't have to be exactly the same as your project(s) as it would be quite hard sometimes. A form and a single component should be more than enough.

Thank you

**Comment by spidertwin2** [comment]
Hi @thanh-nguyen-95, thanks for pointing that out..sorry for the code dump!

Please see [this](https://codesandbox.io/s/heuristic-matsumoto-oiti4?file=/src/index.js) sandbox.

:beach_umbrella:

**Comment by thanh-nguyen-95** [comment, marked as answer]
Here's a working version of your code: https://codesandbox.io/s/black-haze-7o7gz?file=/src/AutoTextarea.tsx

Since your `AutoTextarea` is controlled by both `currentValue` and `react-hook-form`'s value, you must "sync" both of them for it to work. What I did were updating the reset `useEffect`, updating `handleChange` and removing `value` props from `TextArea`.

I left some comments explaining what I did in the code. If you have any further question, feel free to ask.

**Reply by spidertwin2** [reply]
how brilliant! thanks

:pray:

**Reply by bluebill1049** [reply]
thank you @thanh-nguyen-95
